### ISSUE TITLE
WiX v4 compression does not create files as small as v3

### ISSUE THREAD

**Issue by Mertsch** [wixtoolset/issues#7101]
## Bugs

If this issue is a bug:

* Which version of WiX are you building with?

> 4.0.0-rc.1

* Which version of Visual Studio are you building with (if any)?

> 17.4.3

* Which version of the WiX Toolset Visual Studio Extension are you building with (if any)?

> 0.9.1.1

* Which version of .NET are you building with?

> .NET 6

* If the problem occurs when installing your packages built with WiX, what is the version of Windows the package is running on?

> n/a

* Describe the problem and the steps to reproduce it.

> When building an MSI with WiX v3, the .cab file with high compression is roughly 10% smaller then with v4. Even though the files and compression level is the same.
When inspecting the binary a pattern shows.
v3 has an `yyyyyyy` block, then starts with the files (names)
![v3](https://user-images.githubusercontent.com/9402861/208300963-debd04ba-347c-4d29-8840-7e33dc0d610d.png)

v4 has an `yyyyyyy` block, then starts with a lot of repeating patterns that are not present in v3 anywhere
![v4](https://user-images.githubusercontent.com/9402861/208300968-35a9dac6-6f63-4bc7-a613-55066ab58887.png)

only much later after this weirdly suspicious looking block does the file block start
![v4_1](https://user-images.githubusercontent.com/9402861/208301102-9cbd663f-93ff-457b-9d7a-032c15cbc9cf.png)

* Describe the behavior you expected and how it differed from the actual behavior.

> v4 should compress as efficient as v3 or better

**Comment by barnson** [comment]
How are you setting the compression level?

**Comment by Mertsch** [comment]
Yesh, to high. But normal shows the same behavior, just a bit large. I confirmed that the compression level is actually applied by opening the .cab with 7-Zip and in both cases (normal/high) they use the same compression (LZX:21), but have different sizes.

**Comment by barnson** [comment]
How did you author the compression level? I'm getting expected results from a simple test:

534,904,385  0none.cab
154,716,200  1low.cab
130,857,677  2medium.cab
111,988,095  3high.cab
162,193,670  4mszip.cab

**Comment by Mertsch** [comment]
With `MediaTemplate`
```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
  <?include Product.Details.wxi?>

  <Package Name="$(var.ProductDisplayName) $(var.InformationalVersion)"
           Language="!(loc.LanguageID)"
           Version="$(var.Version)"
           Manufacturer="$(var.CompanyName)"
           UpgradeCode="$(var.UpgradeCode)"
           Scope="perMachine">

    <MajorUpgrade DowngradeErrorMessage="!(loc.DowngradeErrorMessage)" AllowSameVersionUpgrades="yes"/>
    <MediaTemplate EmbedCab="yes" CompressionLevel="high"/>
…
      <ComponentGroupRef Id="ComponentGroup.Generated"/>
    </Feature>
  </Package>
</Wix>
```

As said, the difference is between v3 and v4, within a version the compression behaves as you say (reasonable). just Wix v4 has ~2 MB more (my Setup is ~28MB)

**Comment by barnson** [comment]
I get identical results in v3 and v4. Can you provide a repo with a repro project?

**Comment by robmen** [comment]
One thing to note is that WiX v4 changed the order of files added to the cabinet to improve extraction time. It is _possible_ the new order does not compress certain content as well as v3. However, that isn't clear at this moment, and the ability to reproduce the issue and continue optimizing cabinet layout would be interesting.

**Comment by chrpai** [comment]
TLDR?   Summary: WiXUI adds a tiny bit of bloat but cab compression is somehow much better in v4.

I just did some testing.  I created two projects, authored them the same way and upgraded one of them to v4.  The only source of bloat that I detected was WiXUI which is slightly more svelt in v4.   Using my templates as the starting point, the v3 MSI was 236kb and the v4 MSI was 336kb.  Looking at the WiXUICA DLL the v3 was 113kb and the v4 was 145kb.  I don't know where the other 68kb is hiding.   If I remove the UI both v3 and v4 create an 80kb MSI.

Testing cab compression, I arbitrarily picked my Python311 directory which is 6,161 files and 140mb of content.   Using high compression for both v4 created a much smaller MSI.  Weighing in at 30.3MB vs wix v3 which was 34.1MB.  I did admin installs of both and the  extracted files were all in order for both.   The admin MSI were the same sizes also at 1.03MB.  Normally this would be smaller but my authoring tool is quite chatty when generating primary keys.

**Comment by robmen** [comment]
@Mertsch please build your installation packages with the cabinets external.

Then, provide the absolute sizes of the v3 MSI and v3 cabinet(s) and the v4 MSI and v4 cabinet(s). The absolute numbers will allow us to better reason about the differences in sizes.

**Comment by Mertsch** [comment]
@robmen I have already tried external cab and the sizes are no different from .msi, as the .msi is headless (bundled in a Bundle).
I will get back with absolute numbers...

**Comment by barnson** [comment]
We expect minor differences from things like custom action binary sizes (because it's a different compiler). But compression itself looks to be correct.
